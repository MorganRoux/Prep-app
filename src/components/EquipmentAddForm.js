import React from 'react'
import { Paper, TableBody, TableRow, TableCell, TableFooter } from '@material-ui/core';
import IconButton from '@material-ui/core/IconButton';
import Icon from '@material-ui/core/Icon';

import { startAddEquipment } from '../actions/equipment'

import { connect } from 'react-redux'
import { withSnackbar } from 'notistack'

export class EquipmentAddForm extends React.Component {
    
    constructor(props) {
        super(props);

        this.state = {
            quantity: '',
            publicName: ''
        }
    }

    handleCloseSnackBar = (event, reason) => {
        if (reason === 'clickaway') {
          return;
        }
    
        this.setState({ snackBarOpen: false });
      };

    onQuantityChange = (e) => {
        const quantity = e.target.value;
        if(quantity.match(/^\d*$/)) {
            this.setState( () => ({quantity}));
        }
    }

    onPublicNameChange = (e) => {
        const publicName = e.target.value;
        this.setState( () => ({publicName}));
    }

    onAdd = (e) => {
        e.preventDefault()
        // test the inputs
        if(!this.state.publicName || !this.state.quantity) {
            this.props.enqueueSnackbar('Informations erronnées', {variant:'error'});
            return;
        }

        // search the item in the stocklist
        const itemToAdd = this.props.stocklist.find( (item) => {
            return (item.publicName === this.state.publicName)
        });
        if (!itemToAdd) { 
            this.props.enqueueSnackbar('Élement introuvable', {variant:'error'});
            return;
        }

        // add item
        this.props.startAddEquipment({
            ...itemToAdd,
            quantity: this.state.quantity
        });
        const s = this.state.quantity === '1' ? '':'s'
        this.props.enqueueSnackbar(
            `${this.state.quantity}x ${this.state.publicName} ajouté${s}`, {variant:'success'});

        this.setState( () => ({
            quantity: '',
            publicName:''
        }));
    }

    render () {
        return (
            <TableFooter>
                <TableRow className = "addForm-row">
                    <TableCell />
                    <TableCell colSpan="4">
                        <form onSubmit={this.onAdd}>
                        <input 
                        type="text"
                        name= "quantity"
                        autoFocus
                        placeholder=""
                        value={this.state.quantity}
                        onChange = {this.onQuantityChange}
                        />
                 
                    
                        <input 
                        type="text"
                        name= "publicName"
                        placeholder=""
                        value={this.state.publicName}
                        onChange = {this.onPublicNameChange}
                        />
                    
                        <IconButton className="add-button" aria-label="Add" type="submit">
                            <Icon color="primary">
                            add_circle
                            </Icon>
                        </IconButton>
                        </form>
                    </TableCell>
                </TableRow>
                </TableFooter>
        );
    }
}

const mapStateToProps = (state) => {
    return {
        equipements: state.equipements,
        stocklist: state.stocklist,
    }
}

const mapDispatchToProps = (dispatch) => {
    return {
        startAddEquipment: (item) => dispatch(startAddEquipment(item))
    }
}

export default connect(mapStateToProps, mapDispatchToProps)(withSnackbar(EquipmentAddForm))