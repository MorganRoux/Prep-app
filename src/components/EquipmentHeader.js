import React from 'react'
import { connect } from 'react-redux';
import InputBase from '@material-ui/core/InputBase';
import IconButton from '@material-ui/core/IconButton';
import MenuIcon from '@material-ui/icons/Menu';
import SearchIcon from '@material-ui/icons/Search';
import { setTextFilter } from '../actions/filters';
import { startLogout } from '../actions/user';

export class EquipmentHeader extends React.Component {

    constructor (props) {
        super(props);
        this.state = {
            searchText : ''
        }
    }

    onSearchChange = (e) => {

        const searchText = e.target.value;
        this.setState(() => ({searchText}));
        this.props.setTextFilter(searchText);
    }
    
    render () {
        return (
            <div style = {{display: 'flex', justifyContent: 'space-between'}}>
                <div style = {{display: 'flex', justifyContent: 'flex-start'}}>
                    <IconButton>
                        <MenuIcon />
                    </IconButton>
                    <h1>Equipment List</h1>
                </div>
                <div style = {{display: 'flex', justifyContent: 'flex-end'}}>
                    <InputBase 
                        className="search-field"
                        placeholder="Rechercher" 
                        value={this.state.searchText}
                        onChange={this.onSearchChange}
                    />
                    <IconButton>
                        <SearchIcon />
                    </IconButton>
                    <button className = "logout-button" onClick={this.props.startLogout}>Logout</button>
                </div>
            </div>
    )};

}

const mapDispatchToProps = (dispatch) => ({
    setTextFilter : (text) =>  dispatch(setTextFilter(text)),
    startLogout : () => dispatch(startLogout())
});

export default connect(null, mapDispatchToProps)(EquipmentHeader)