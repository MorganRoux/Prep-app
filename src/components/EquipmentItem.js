import React from 'react'
import { connect } from 'react-redux';
import {TableBody, TableRow, TableCell } from '@material-ui/core';
import { removeEquipment } from '../actions/equipment';
import IconButton from '@material-ui/core/IconButton';
import DeleteIcon from '@material-ui/icons/Delete';
import Checkbox from '@material-ui/core/Checkbox';
import { withSnackbar } from 'notistack';
import { SortableElement } from 'react-sortable-hoc';

export class EquipmentItem extends React.Component {
    constructor(props) {
        super(props);
        const stockItem = props.stocklist.find((item) => {
            return ( item.stockName === props.equipment.stockName )
        });
        this.state = {
            stockName: props.equipment.stockName,
            id: props.equipment.id,
            quantity: props.equipment.quantity,
            stockItem: stockItem
        }
        
    }
    
    onRemove = () => {
        this.props.removeEquipment(this.state.id);
        this.props.enqueueSnackbar('Item supprimÃ©', {variant:'success'});
    }
    
    onQuantityChange = () => {

    }

    onPublicNameChange = () => {
        console.log('change !');
    }

    onStockNameChange = () => {

    }
    
    render() {
        return (
            <TableRow key={`${this.state.id}-row`}>
                <TableCell><Checkbox /></TableCell>
                <TableCell key = {`${this.state.id}-cell-1`}>{this.state.quantity}</TableCell>
                <TableCell key = {`${this.state.id}-cell-2`}>{this.state.stockItem.publicName}</TableCell>
                <TableCell key = {`${this.state.id}-cell-3`}>{this.state.stockItem.stockName}</TableCell>
                <TableCell key = {`${this.state.id}-cell-4`}>
                    <IconButton 
                        className = {'delbutton'} 
                        aria-label="Delete" 
                        onClick = {this.onRemove}>
                    <DeleteIcon />
                    </IconButton>
                </TableCell>   
            </TableRow>
        );
    };
}

const mapStateToProps = (state, props) => {
    return {
        equipments: state.equipments,
        filters: state.filters,
        stocklist : state.stocklist
    };
} 

const mapDispatchToProps = (dispatch, props) => {
    return {
        removeEquipment: (id) => dispatch(removeEquipment(id)),
        editEquipment: () => dispatch(editEquipment()),
        addEquipment: () => dispatch(addEquipment())
    };
}

const EquipmentItemWithBody = (props) => (
    <TableBody>
        <EquipmentItem {...props} />
    </TableBody>
);

// export default connect(mapStateToProps,mapDispatchToProps)(withSnackbar(EquipmentItem));

export const EquipmentKitItem = connect(mapStateToProps,mapDispatchToProps)(withSnackbar(EquipmentItem));
export const EquipmentListItemSortable = connect(mapStateToProps,mapDispatchToProps)(withSnackbar(SortableElement(EquipmentItemWithBody)));
