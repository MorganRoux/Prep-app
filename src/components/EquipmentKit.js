import uuid from 'uuid'
import React from 'react';
import EquipmentListItem from './EquipmentListItem';
import { connect } from 'react-redux';
import { removeEquipmentÂ } from '../actions/equipment';

import {TableBody, TableRow, TableCell } from '@material-ui/core';
export class EquipmentKit extends React.Component{
    constructor(props) {
        super(props);
    }

    onRemove = (key) => {
        this.props.removeEquipment(key);
    }

    render() {
        return (
            <TableBody>
                <TableRow><TableCell colSpan="4">EquipmentKit</TableCell></TableRow>
                {   this.props.equipments.map((equipment) => {
                    return (
                    equipment.parentKey === this.props.kit.key && (
                        <EquipmentListItem 
                        key={uuid()}
                        equipment={equipment}
                        onRemove={this.onRemove}
                        />
                    )
                )})}
            </TableBody>
        );
    }
}

const mapStateToProps = (state) => {
    return {
         kits: state.kits,
         equipments: state.equipments
    };
}

const mapDispatchToProps = (dispatch) => {
    return {
        removeEquipment: (key) => dispatch(removeEquipment(key))
    }
}

export default connect(mapStateToProps, mapDispatchToProps)(EquipmentKit)