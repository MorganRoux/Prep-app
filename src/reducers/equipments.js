
const equipmentReducerDefaultState = [{
    category: 'microphone',
    brand: 'shure',
    description:'micro cardioide',
    quantity: 1,
    publicName: 'sm57',
    stockName: 'sm57A',
    id: 1
}, {
    category: 'microphone',
    brand: 'shure',
    description:'micro grosse caisse',
    quantity: 1,
    publicName: 'Beta91',
    stockName: 'B91',
    id: 2
}, {
    category: 'microphone',
    brand: 'shure',
    description:'micro grosse caisse',
    quantity: 5,
    publicName: 'Beta52',
    stockName: 'B52',
    id: 3
}];

const equipmentsReducer = (state = equipmentReducerDefaultState, action) => {
    switch(action.type){

    case 'REMOVE_EQUIPMENT' :
        const newState = state.filter((equipment) => (action.id !== equipment.id) );
        return newState;
    
    case 'EDIT_EQUIPMENT' :
    case 'ADD_EQUIPMENT' :
    default:
        return state;
    }
}

export default equipmentsReducer