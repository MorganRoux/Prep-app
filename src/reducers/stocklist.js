import stocklist from '../tests/fixtures/stocklist';
const stocklistReducerDefaultState = stocklist;

const stocklistReducer = (state = stocklistReducerDefaultState, action) => {
    switch(action.type){

    default:
        return state;
    }
}

export default stocklistReducer