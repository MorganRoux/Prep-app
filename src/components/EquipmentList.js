import React from 'react';
import EquipmentListItem from './EquipmentListItem';
import { connect } from 'react-redux'
import EquipmentKit from './EquipmentKit';

export const EquipmentList = (props) => (
    <div>
        <h3>EquipmentList</h3>
        
        { props.equipments.map((equipment) => {
            
            return (equipment.category === 'kit') ? (
                <EquipmentKit />
            ):(
                <EquipmentListItem 
                key={equipment.id}
                equipment = {equipment}
                /> 
            ); 
        })}
    </div>
);

const mapStateToProps = (state) => {
    return {
        equipments: state.equipments
    }
};

export default connect(mapStateToProps)(EquipmentList)