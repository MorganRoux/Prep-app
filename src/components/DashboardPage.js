import React from 'react';
import EquipmentList from './EquipmentList';

const DashboardPage = () => {
    return (
        <div>
            <h1>Dashboard Page</h1>
            <EquipmentList />
        </div>
        
    );
};

export default DashboardPage;