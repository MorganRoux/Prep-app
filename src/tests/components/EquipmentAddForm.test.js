import React from 'react'
import { EquipmentAddForm } from '../../components/EquipmentAddForm';
import { shallow } from 'enzyme';
import equipments from '../fixtures/equipment';
import stocklist from '../fixtures/stocklist';


let wrapper, startAddEquipment, enqueueSnackbar;

beforeEach(() => {
    startAddEquipment = jest.fn();
    enqueueSnackbar = jest.fn();
    wrapper = shallow(<EquipmentAddForm 
        kit = {equipments[3]}
        startAddEquipment = {startAddEquipment}
        enqueueSnackbar = {enqueueSnackbar}
        equipments= {equipments}
        stocklist = {stocklist}
    />);

});

test('should render EquipmentAddForm', () => {
    
     expect(wrapper).toMatchSnapshot();
 });

 test('should handle on quantity change', () => {
     const value = '3';
     wrapper.find('input').at(0).simulate('change', { 
         target: {value }
     });
     expect(wrapper.state('quantity')).toBe(value);

 });

 test('should not change if quantity is not a number', () => {

    const value = '3.5';
     wrapper.find('input').at(0).simulate('change', { 
         target: {value }
     });
     expect(wrapper.state('quantity')).toBe('');
    
})
 test('should handle on publicName change', () => {
    const value = 'name';
    wrapper.find('input').at(1).simulate('change', { 
        target: {value }
    });
    expect(wrapper.state('publicName')).toBe(value);
});

test('OnAdd : should display error message if fields empty on add', () => {
    wrapper.find('form').simulate('submit',{preventDefault: () => {}});
    expect(enqueueSnackbar).toHaveBeenLastCalledWith('Informations erronnées', {variant:'error'});
});

test('OnAdd : should display error message if item not found', () => {
    const quantity = '3';
    const publicName = 'notfounditem';

    wrapper.setState(()=>({quantity, publicName}));
    wrapper.find('form').simulate('submit', {preventDefault: () => {}});
    expect(enqueueSnackbar).toHaveBeenLastCalledWith('Élement introuvable', {variant:'error'});
});

test('onAdd : should add an item', () => {
    const quantity = '3';
    const arg = {...stocklist[0], quantity};

    const newState = {
        quantity: '',
        publicName:'', 
    }
    wrapper.setState(()=>({quantity, publicName: stocklist[0].publicName}));
    wrapper.find('form').simulate('submit',{preventDefault: () => {}});

    expect(startAddEquipment).toHaveBeenCalledWith(arg);
    expect(wrapper.state()).toEqual(newState);
    expect(enqueueSnackbar).toHaveBeenLastCalledWith('3x sm57 ajoutés', {variant:'success'});
});



