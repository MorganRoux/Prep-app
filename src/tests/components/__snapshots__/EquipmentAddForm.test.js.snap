// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`should render EquipmentAddForm 1`] = `
<WithStyles(TableFooter)>
  <WithStyles(TableRow)
    className="addForm-row"
  >
    <WithStyles(TableCell) />
    <WithStyles(TableCell)
      colSpan="4"
    >
      <form
        onSubmit={[Function]}
      >
        <input
          autoFocus={true}
          name="quantity"
          onChange={[Function]}
          placeholder=""
          type="text"
          value=""
        />
        <input
          name="publicName"
          onChange={[Function]}
          placeholder=""
          type="text"
          value=""
        />
        <WithStyles(IconButton)
          aria-label="Add"
          className="add-button"
          type="submit"
        >
          <WithStyles(Icon)
            color="primary"
          >
            add_circle
          </WithStyles(Icon)>
        </WithStyles(IconButton)>
      </form>
    </WithStyles(TableCell)>
  </WithStyles(TableRow)>
</WithStyles(TableFooter)>
`;
